src/object.cr

```diff
@@ -178,833 +178,841 @@ class Object
-  #
-  #   def name=(@name)
-  #   end
-  # end
-  # ```
-  macro setter(*names)
-    {% for name in names %}
-      {% if name.is_a?(TypeDeclaration) %}
-        @{{name}}
-
-        def {{name.var.id}}=(@{{name.var.id}} : {{name.type}})
-        end
-      {% elsif name.is_a?(Assign) %}
-        @{{name}}
-
-        def {{name.target.id}}=(@{{name.target.id}})
-        end
-      {% else %}
-        def {{name.id}}=(@{{name.id}})
-        end
-      {% end %}
-    {% end %}
-  end
-
-  # Defines property methods for each of the given arguments.
-  #
-  # Writing:
-  #
-  # ```
-  # class Person
-  #   property name
-  # end
-  # ```
-  #
-  # Is the same as writing:
-  #
-  # ```
-  # class Person
-  #   def name=(@name)
-  #   end
-  #
-  #   def name
-  #     @name
-  #   end
-  # end
-  # ```
-  #
-  # The arguments can be string literals, symbol literals or plain names:
-  #
-  # ```
-  # class Person
-  #   property :name, "age"
-  # end
-  # ```
-  #
-  # If a type declaration is given, an instance variable with that name
-  # is declared with that type.
-  #
-  # ```
-  # class Person
-  #   property name : String
-  # end
-  # ```
-  #
-  # Is the same as writing:
-  #
-  # ```
-  # class Person
-  #   @name : String
-  #
-  #   def name=(@name)
-  #   end
-  #
-  #   def name
-  #     @name
-  #   end
-  # end
-  # ```
-  #
-  # The type declaration can also include an initial value:
-  #
-  # ```
-  # class Person
-  #   property name : String = "John Doe"
-  # end
-  # ```
-  #
-  # Is the same as writing:
-  #
-  # ```
-  # class Person
-  #   @name : String = "John Doe"
-  #
-  #   def name=(@name : String)
-  #   end
-  #
-  #   def name
-  #     @name
-  #   end
-  # end
-  # ```
-  #
-  # An assignment can be passed too, but in this case the type of the
-  # instance variable must be easily inferrable from the initial value:
-  #
-  # ```
-  # class Person
-  #   property name = "John Doe"
-  # end
-  # ```
-  #
-  # Is the same as writing:
-  #
-  # ```
-  # class Person
-  #   @name = "John Doe"
-  #
-  #   def name=(@name : String)
-  #   end
-  #
-  #   def name
-  #     @name
-  #   end
-  # end
-  # ```
-  #
-  # If a block is given to the macro, a property is generated
-  # with an instance variable that is lazily initialized with
-  # the block's contents:
-  #
-  # ```
-  # class Person
-  #   property(birth_date) { Time.now }
-  # end
-  # ```
-  #
-  # Is the same as writing:
-  #
-  # ```
-  # class Person
-  #   def birth_date
-  #     @birth_date ||= Time.now
-  #   end
-  #
-  #   def birth_date=(@birth_date)
-  #   end
-  # end
-  # ```
-  macro property(*names, &block)
-    {% if block %}
-      {% if names.size != 1 %}
-        {{ raise "only one argument can be passed to `property` with a block" }}
-      {% end %}
+  {% for prefixes in { {"", "", "@"}, {"class_", "self.", "@@"} } %}
+    {%
+      macro_prefix = prefixes[0]
+      method_prefix = prefixes[1]
+      var_prefix = prefixes[2]
```

Comment:
@Sija We can do this:

```cr
macro_prefix = prefixes[0].id
method_prefix = prefixes[1].id
var_prefix = prefixes[2].id
```

Then everywhere below we don't need to use `.id` and code is shorter and clearer (and slightly faster). What do you think?